### DISCUSSION TITLE
Need to connect to transmitter before we can start sensor

### DISCUSSION THREAD

**Discussion by old-square-eyes** [NightscoutFoundation/xDrip#3078] [Bug reports] [answered]
After 20230328 a change was rolled out to make sure there was a valid timestamp for transmitter days resulting in "Need to connect to transmitter before we can start sensor". I believe this timestamp happens shortly after a TX is activated. And the rationale was to make sure a TX was ready to take a sensor start command (and importantly avoid potential "stuck" commands in the xDrip TX queue).

While I never suffered from a stuck command in my life... I seem to reliably get blocked from starting a sensor by this countermeasure, permanently. I have to roll back to 20230328 every time I change my sensor as a workaround.

I have an Anubis and I factory reset it after every use (have two Anubis and alternate to overlap, sleeping the outgoing TX to save battery). For whatever reason TX days stays at NaN (or "/6214") until after I start it, making this a race condition in my situation. I can't start because of the block, and I can't get unblocked because of the start limitation.

I know my TX connects to my phone because it takes a battery reading and the "Last Connected" seconds rolls over. Yet because the timestamp doesn't update I'm check mated.

[Original FB post](https://www.facebook.com/groups/xDripG5/posts/9057191481020933/)

![Screenshot_20230815_165244_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/25760491/6f1b8792-d9d9-4995-8814-5ba499bb7b50)

**Comment by Navid200** [comment]
After you start by rolling back, how long does it take for transmitter days to show 0?
Is it 15 minutes, half an hour, an hour?

Will it not do the same if you don't roll back?
I mean if it takes half an hour for transmitter days to show0, will it not do that if you don't roll back?
Or will it never change to 0 unless you roll back and start?

**Comment by Navid200** [comment]
I just recreated the problem by resetting the transmitter without removing the pairing from the phone.
After forgetting the device and approving the resulting pair request that followed, it took 30 minutes for the transmitter days to show 0.

Let me run another test.

**Comment by Navid200** [comment]
This time, I didn't use forget device.
After 44 minutes, there was still no way to start.

This time, I enabled `Allow OB1 unbonding`.  This setting is recommended to be left disabled.
After doing that, it unpaired and I got a pair request.
After 14 minutes, I was able to start.

**Reply by Navid200** [reply]
I'm running a third test.

**Reply by Navid200** [reply]
This time, after resetting the transmitter, I forgot device in xDrip before activating the transmitter.
My test phone is a Motorola Android 11 phone.  It suffers from occasional disconnects.  As it happens, after the pair request and the first successful read cycle, it disconnected.  I restarted the phone.  After the next read cycle, it allowed me to start.

**Comment by Navid200** [comment]
This is my understanding of what happens.

Bluetooth pairing is a 2-way process.  When there is a pairing established (pair request approved), xDrip stops looking for a transmitter to pair with.
After you reset the transmitter, the transmitter has no idea about the phone because you reset it.  It needs to pair before it can do anything.   So, it's like there is a one-sided pairing where xDrip assumes that it is linked with the transmitter while the transmitter does not.  So, there is never a correct connectivity established.

The allow OB1 unbonding, I believe, was put in place for exactly this situation.  But, it caused more harm than help.  So, we are telling everyone to disable it.

The way I see it, you can enable allow OB1 unbonding if it does not cause frequent pair requests for you.
Alternatively, you can forget device as soon as you reset your transmitter to avoid a one-sided pairing.

I can see that others may get locked up the way you were using recommended settings.  
Thanks for reporting this.

Please let me discuss this with others before we can see what we need to do next.
But, in the meantime, if the two options I have suggested fail, please let me know.

**Reply by old-square-eyes** [reply]
Thanks for looking into this. I had tried various manual unbonding already. Before I reset at the end of a sensor (with the Anubis tool), I always forget device with xdrip (and disable collector). I will double check next time post Anubis tool (in android OS BT paired devices since Anubis tool has no such feature).

What if we flip it on its head? Why not use a battery read as the gate for your readiness check?

I'll test your other requests too. Other thing to note is I swear this only happens with one of my Anubis. Could be a coincidence but I'll keep an eye on it.

**Reply by Navid200** [reply]
I don't think it is going to change anything if we switch to battery.
The lock scenario you have experienced is a situation in which, there is no connectivity between xDrip and the transmitter.  Even though the status page shows last connected to be less than 5 minutes, that's only when xDrip was able to see the ping from the transmitter.  But, as far as the transmitter is concerned, there is no link.  So, xDrip cannot read anything, neither transmitter days nor battery voltages.

**Reply by Navid200** [reply]
If that's not the case, meaning if xDrip can indeed read the battery voltages without a pairing in place, then voltages being present is not a valid representation that we have a link.
If we don't have a link, starting the sensor could cause a start command getting stuck in the queue.

**Comment by old-square-eyes** [comment]
My screen shot has a battery reading but no transmitter days. My hunch is a
connection has occurred and it's G2G. Hence when I use an older version
without the check, I progress without issue.

I don't fully understand the other stuck command issue though. I'm guessing
I'm vulnerable to it somehow.

Maybe don't have a hard gate, but instead, auto clear any stuck commands
after a threshold and prompt user start has failed?

**Reply by old-square-eyes** [reply]
Making people wait on every sensor change (while good intentions) is quite a large inconvenience e.g. night time replacements. I'd rather chance my luck with a stuck interaction and try again in the rare occurrence that it happens.

**Comment by old-square-eyes** [comment]
@Navid200 Going through a changeover now. Tried each of the below, waiting for 15-20 minutes between trying each.

- Allow unbonding is enabled
- Also tried "Special Pairing Workaround"
- Unpaired in Android, repaired when I got the popup 10mins later
- Forgot device in xDrip/restart collector
- Tried unbond before each read

As before I get a battery read so I'm convinced a connection happens. It just feels like a race condition with this Anubis.
<img width="356" alt="image" src="https://github.com/NightscoutFoundation/xDrip/assets/25760491/cf0119d9-0375-420f-a07f-590157bf61a2">
<img width="341" alt="image" src="https://github.com/NightscoutFoundation/xDrip/assets/25760491/57464d11-cfcb-4826-9805-eaa6a5a5568b">
<img width="347" alt="image" src="https://github.com/NightscoutFoundation/xDrip/assets/25760491/ca2b440e-f7c8-4886-8b39-8e5c46106f6d">

**Comment by old-square-eyes** [comment]
Rolled back versions and it started instantly (without the gate obviously) but as soon as the start command was issued the TX timestamp populated in status with a 0 days.

@Navid200 any chance you can bury a setting to disable the gate so I can be done with this? Been slogging away all year :) Changeovers are stressful enough.

**Reply by Navid200** [reply]
When I recreated this problem, after I forgot device, I got a pair request on the next read cycle.  After the pair request was approved, waiting for one more read cycle, I was able to start sensor even though transmitter days was still showing blank.

The last change I made to this behavior was on June 24, 2023.  https://github.com/NightscoutFoundation/xDrip/pull/2757
Even that became part of the beta (stable) release on July 16, 2023.

But, the block on start when transmitter days is unknown was actually added on April 11, 2023.  https://github.com/NightscoutFoundation/xDrip/pull/2746  So, it became part of the stable release on  May 3, 2023.

If this was a problem with xDrip, causing people to get locked, we should have been getting reports from others by now.
I'm not sure why you experience this with one of your two transmitters and not the other.
Have you seen anyone else experiencing this problem in any forum?

I am asking other developers to look at this (thanks for reporting) and advise what we should do.

**Reply by secabeen** [reply]
I saw a similar situation on my Anubis when starting up after a factory deep-sleep.  Like Navid, I regularly factory sleep my transmitter before starting it up again after some weeks.  I think most users just stick with a single transmitter, replacing the battery as needed, but not using the factory sleep functionality.

In my case, I installed the factory-slept transmitter, paired successfully, but then immediately lost pairing and data acquisition due to a battery level report of '0'.  I didn't check the transmitter days, but I would expect it was blank like Navid.  In my case, I was able to get started by completely removing and re-pairing the transmitter after waiting 30 minutes for the battery data to show up, but I expect there is some weirdness in how Anubis behaves regarding transmitter days and battery levels after factory sleep wakeup.

**Comment by old-square-eyes** [comment]
Discussion on the Anubis page and I have tagged the dev who chimed in https://www.facebook.com/groups/anubisg6/posts/843692667413110/

**Reply by Navid200** [reply]
I don't know why your two transmitters behave differently considering the firmwares are identical.
Is it possible that there is another device, like a watch, that may be linked with only one of the two transmitters?

I have not been able to recreate the problem you describe where forgetting the device and approving the pair request does not resolve the problem.

I know that if we change xDrip, we can help you with one of your two transmitters.
But, we cannot change xDrip just for a single transmitter.

We are not getting start commands getting stuck in the queue any longer.  So, the change we made 5 months ago is doing its job.   I don't want those problems to come back.

**Reply by old-square-eyes** [reply]
I'll try to reproduce with my other Anubis this weekend. Since the problematic one is now in I won't be able to tinker with that for a week. I do have a watch, but struggling to think how it could cause ongoing issues across many resets and starts over months.

**Reply by old-square-eyes** [reply]
@Navid200 would you please be able to source the Variant 1 20230328 apk for testing since it has aged off the repo? Come to think of it... Running main + variant may be a contributing factor so will keep an eye on that. I do housekeep properly though by forgetting devices and disabling collection at appropriate times. ie never have both apps competing for the same TX.

**Reply by Navid200** [reply]
I don't have a backup for variants, sorry.
You will have to use the variants from the stable release before that:  https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.11.16

**Reply by Navid200** [reply]
Would you happen to have an old phone in the house for a test?
The next time this happens, please use the old phone and install the regular current stable release on it, disable collection on your main phone, use the old phone and see if it connects to the transmitter you want to start.

If it starts, you can then disable collection on the old phone and the regular phone should then be able to connect I hope.

**Reply by old-square-eyes** [reply]
@Navid200 yes I can try this next time with my old S8

**Reply by Navid200** [reply]
This is what I suggest assuming you overlap one day:

Say you have TX1 on your main phone.
When day 10 of the sensor starts, use your old phone, with the current stable release of xDrip.
Enter the transmitter ID of TX2 on the old phone.  Insert a new sensor on your other arm.  Snap in TX2 into it.
Wait for a pair request on the old phone.  Pair.  Wait for the next read cycle.  You should now be able to start sensor on TX2.
After the sensor starts, wait for 2 more read cycles confirming that everything is fine and the warm-up is shown on the main screen and status page.
Now, you can turn off the old phone.

Next day, the sensor on TX1 ends.
Change the transmitter ID on your main phone to that of TX2.  Wait for it to pair.
In a few read cycles, it should start showing readings.

Peel off TX1 and its sensor from the other arm.

When you get to day 10 on TX2, repeat everything again replacing TX1 and TX2 in the instructions.

**Comment by Navid200** [comment]
I have been able to recreate this problem.
I will need a few days to hopefully find a solution.

In the meantime, you can do what you did, temporarily install an older xDrip to start.  
Thanks for reporting this.

**Comment by Navid200** [comment]
xDrip has a persistent storage that contains information about a transmitter.
It uses the data from that storage instead of requesting all of it every 5 minutes from the transmitter.

After a transmitter is hard reset, the timestamp of the transmitter goes back to 0.  If you perform a hard reset using xDrip, the persistent store data is also flushed from the transmitter data.  As a result, on the next contact, xDrip will ask for all the data from the transmitter and loads the persistent store with updated values.

If you use a different app to hard reset a transmitter, the data in the xDrip persistent store (the timestamp) will be incorrect.
This is why you could experience a lock for a while until xDrip gets an update from the transmitter.

If you always use xDrip to perform a hard reset, you will never experience the lock situation.
Even if you use another app and experience the lock, you can perform another hard reset using xDrip to remedy that.

**Reply by secabeen** [reply]
This is helpful context.  Perhaps then what we need a documentation update that describes how to use xDrip alone to do an Anubis factory reset and deep sleep.  Alternately, an option/workflow for anubis users who have done the factory reset/deep sleep in AnubisT00L to clear the related storage in xDrip.

**Reply by Navid200** [reply]
This is how xDrip hard reset works:
https://navid200.github.io/xDrip/docs/Hard-Reset.html

**Reply by secabeen** [reply]
Right, this looks like the page that could use some updating.  For example, it currently says this "You can also hard reset a G6, using xDrip, if the Firmware version is 1.6.5.25 or prior to that. At this point, there is no way to hard reset any other one of G6 transmitters."  Perhaps this section could mention that for modified transmitters, G6 resets are still possible.

Later it says "Before you can hard reset a transmitter using xDrip, you need to have proper connectivity to it."  That implies that you could not use this function to clear the transmitter storage in xDrip for a transmitter you have already reset with another app.

**Reply by old-square-eyes** [reply]
Thanks for digging more @Navid200. I'll be sure to test. However, I recall in my desperation one time I did try running the hard reset from xdrip, in addition, to get it to connect. But I'll be sure to double check next time round. My current sensor is the first one to have gotten past 10 days in months but probably going to die soon so will report back.

**Reply by Navid200** [reply]
@secabeen 
There is no way any app, including xDrip, can perform a software hard reset on a transmitter without first connecting to it.
That's what the sentence you have quoted intends to imply.

Your conclusion pursuant to that is not 100% valid.
If you perform a hard reset using another app, xDrip will refuse to start sensor for a while.  
That does not mean that if you perform a hard reset using another app, xDrip will not be able to connect to it.

**Reply by old-square-eyes** [reply]
@Navid200 were you talking to me? I thought you implied I could try that to clear the persistent storage.

**Reply by Navid200** [reply]
@old-square-eyes No, I was not talking to you.  You can see I tagged someone else.

**Reply by secabeen** [reply]
@Navid200 
>  That's what the sentence you have quoted intends to imply.

That makes sense.  Is there any other way to clear the transmitter storage other than the hard reset function?

**Reply by Navid200** [reply]
@secabeen 
> Is there any other way to clear the transmitter storage other than the hard reset function?

If you mean any other way to clear the xDrip persistent storage content related to the transmitter, I can think of the following, most of which are destructive!

1- If you clear the cache and storage content related to xDrip in your Android storage settings, that will do it.   But, that will erase the database.  So, that is a big no no.  

2- If you uninstall xDrip and do a fresh install, that will do it.  Again, that will erase the database.  Again, a very bad idea.

3- You can use a temporary phone just for starting the sensor.  After you start the sensor, you use your main phone to get readings and upload to Nightsocut etc.  And you uninstall xDrip on the temp phone.  



None of those is a recommended practice.  The easiest and simplest approach would be to use xDrip to perform the hard reset.

**Reply by secabeen** [reply]
Okay, I'll take this to the developers of the other app that's doing the hard reset and see if they want to update their recommendations.  Thanks!

**Comment by old-square-eyes** [comment]
I am doing a swapover now. Hard Reset Transmitter doesn't help to recover from the stuck state. It just says "Hard reset maybe failed". I have been in the stuck state for over an hour. Hard reset has been attempting for most of that time.

**Reply by Navid200** [reply]
I know, I am sorry.  
I believe this will fix it: https://github.com/NightscoutFoundation/xDrip/pull/3123

**Comment by old-square-eyes** [comment]
@Navid200 I was able to start the transmitter on my old phone, with a newer app version. So that (I guess) confirms it was something left over in local DB storage per your previous comments.

**Comment by Navid200** [comment, marked as answer]
A new xDrip Nightly is supposed to address this: https://github.com/NightscoutFoundation/xDrip/releases/tag/2023.10.17

Please use it, or any after that, to address the problem.

You still need to remove the transmitter from the list of previously paired devices on your phone.  Otherwise, you will see deep sleeping error and experience malfunction.
As long as you remove the hard reset transmitter from the list of previously paired devices, the behavior will be identical to that of a brand new standard transmitter.
So, you will need to approve a pair request.  Then, you will need to wait for one or two consecutive read cycles after which,  you will be able to start sensor.

And you can hard reset using xDrip.  But, you don't have to.  This is addressed regardless of how you hard reset.

**Reply by old-square-eyes** [reply]
Awesome. Can't wait to try it.

**Reply by old-square-eyes** [reply]
Fist changeover went without a hitch. Thank you for seeing this through @Navid200

**Reply by Navid200** [reply]
You are more than welcome.
Thanks for patiently explaining the problem while I was trying to convince you that it was a user error when in fact it wasn't.  I didn't know at the time.  And I am sorry about that.  And thanks for reporting.
